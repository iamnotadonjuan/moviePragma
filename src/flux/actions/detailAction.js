import { INIT_FAVORITE, SAVE_RATING } from '../actionTypes'

const initFavorite = (data) => {
  return async dispatch => {
    const favorite = [{
      star_1: false,
      star_2: false,
      star_3: false,
      star_4: false,
      star_5: false
    }
    ]
    dispatch(_initFavorite(favorite))
  }
}

const _initFavorite = (data) => {
  return {
    type: INIT_FAVORITE,
    payload: { data }
  }
}

const saveRating = (data) => {
  return async dispatch => {

    if (data.count > 1) {
      let i;
      for (i = 1; i <= data.count; i++) {
        if (data.value == true && data.count < 5) {
          for (i = data.count; i <= 5; i++) {
            dataCopy = {
              position: 'star_' + i,
              value: data.value
            }
            dispatch(_saveRating(dataCopy))
          }
        } 

        dataCopy = {
          position: 'star_' + i,
          value: data.value
        }
        dispatch(_saveRating(dataCopy))
      }
    }
    else {
      if (data.value == true) {
        _initStar(dispatch, data)
      } else {
        dispatch(_saveRating(data))
      }
    }
  }
}

const _saveRating = (data) => {
  return {
    type: SAVE_RATING,
    payload: { data }
  }
}

const _initStar = (dispatch,data) => {
  for (i = data.count; i <= 5; i++) {
    dataCopy = {
      position: 'star_' + i,
      value: data.value
    }
    dispatch(_saveRating(dataCopy))
  }
}


export { initFavorite, saveRating }