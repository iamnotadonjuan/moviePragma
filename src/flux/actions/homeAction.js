import { MOVIE_LIST } from '../actionTypes'
import API from '../../services/api'

const listMovie =  (data) => {
  return async dispatch => {
    const List = await API.get(data.limit, '3D', data.url)
    dispatch(_listMovie(List))
  }
}

const _listMovie = (data) => {
  return {
    type: MOVIE_LIST,
    payload: { data }
  }
}


export { listMovie }