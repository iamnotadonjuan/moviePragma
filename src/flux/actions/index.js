export { login, initSession, saveUserInfo } from './loginAction'