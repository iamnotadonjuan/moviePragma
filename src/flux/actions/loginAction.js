import { LOGIN_ACTION, INIT_SESSION } from '../actionTypes'

const login = (data) => {
  return dispatch => {
    dispatch(_login(user))
  }
}

const _login = (data) => {
  return {
    type: LOGIN_ACTION,
    paylod: { data }
  }
}

const initSession = data => {
    return dispatch => {
      dispatch(_initSession(data))
    }
}

const _initSession = (data) => {
  return {
    type: INIT_SESSION,
    payload: { data }
  }
}

export { login, initSession }