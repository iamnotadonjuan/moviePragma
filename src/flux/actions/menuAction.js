import { CLOSE_MENU } from '../actionTypes'

const closeMenu =  (data) => {
  return async dispatch => {
    dispatch(_closeMenu(data))
  }
}

const _closeMenu = (data) => {
  return {
    type: CLOSE_MENU,
    payload: { data }
  }
}


export { closeMenu }