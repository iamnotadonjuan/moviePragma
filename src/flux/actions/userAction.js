import { SAVE_USER } from '../actionTypes'
import {
  AsyncStorage
} from 'react-native'

const saveUserInfo = (data) => {
  return async dispatch => {
    try {
      await AsyncStorage.setItem('currentUser', JSON.stringify(data))
      dispatch(_saveUserInfo(data))
    } catch (error) {
      console.log(error)
    }
  }
}

const retrieveUserInfo = async (data) => {
  try {
    return await AsyncStorage.getItem('currentUser')
  } catch (error) {
    console.log(error)
  }
}

const _saveUserInfo = (data) => {
  return {
    type: SAVE_USER,
    payload: { data }
  }
}

export { saveUserInfo, retrieveUserInfo }
