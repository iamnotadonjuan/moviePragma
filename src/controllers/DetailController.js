import React, { Component } from 'react'
import { connect } from 'react-redux'
import { initFavorite, saveRating, openModal } from '../flux/actions/detailAction'
import DetailView from '../views/detailView'



class Detail extends Component {
  constructor(props) {
    super(props)
    this.state = {
      favorite: [{
        star_1: false,
        star_2: false,
        star_3: false,
        star_4: false,
        star_5: false
      }],
      isOpenModal: false
    }
    this._initFavorites();
  }

  async _initFavorites() {
    await this.props._initFavorite()
  }

  saveFavorite = async (data) => {
    let position = data[0]
    let value = data[1]
    let count = data[2]
    await this.props._saveRating({ position, value, count })
  }

  modalDetail = async (data) => {
    const isOpenModal = this.props.isOpenModal;
    await this.props._openModal({ isOpenModal })
  }
  isCloseModal = async () => {
    const isOpenModal = this.props.isOpenModal;
    await this.props._openModal({isOpenModal})
  }

  render() {
    return (
    <DetailView
    navigation={this.props.navigation}
    isOpenModal={this.props.isOpenModal}
    saveFavorite={this.saveFavorite}
    favorite={this.state.favorite}
    modalDetail={this.modalDetail}
    isCloseModal={this.isCloseModal}
    />
    )
  }
}

const mapStateToProps = state => {
  return {
    favorite: state.detail.favorite,
    isOpenModal: state.detail.isOpenModal
  }
}

const mapDispatchToProps = dispatch => {
  return {
    async _initFavorite(data) {
      await dispatch(initFavorite(data))
    },
    async _saveRating(data) {
      await dispatch(saveRating(data))
    },
    async _openModal(data) {
      await dispatch(openModal(data))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Detail)