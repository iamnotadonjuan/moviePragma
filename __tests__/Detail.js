import 'react-native'
import React from 'react'
import DetailController from '../src/controllers/DetailController';
import Adapter from 'enzyme-adapter-react-16'
import { shallow, configure } from 'enzyme'
import thunk from "redux-thunk";
import configureStore from 'redux-mock-store'
import { initFavorite } from '../src/flux/actions/DetailAction'


const buildStore = configureStore([thunk]);


configure({ adapter: new Adapter() })
describe('Testing Menu component', () => {
    it('Menu renderers', () => {
        const wrapper = shallow(
            <DetailController />
        )
        expect(wrapper).toMatchSnapshot()
    })
})

describe('Testing requests', () => {
    let store;
    const initialState = {
        star_1: false,
        star_2: false,
        star_3: false,
        star_4: false,
        star_5: false
    }
    beforeEach(() => {
        store = buildStore(initialState);
    });
    it("init favorite", async () => {
        const favorite = [{
            star_1: false,
            star_2: false,
            star_3: false,
            star_4: false,
            star_5: false
        }
        ]
        await store.dispatch(initFavorite({ favorite }))
        expect(store.getActions()).toMatchSnapshot()
    });
    it("save rating", async () => {
        let position = 'star_2'
        let value = true
        let count = 2
        await store.dispatch(initFavorite({ position, value, count }))
        expect(store.getActions()).toMatchSnapshot()
    });
}
)

