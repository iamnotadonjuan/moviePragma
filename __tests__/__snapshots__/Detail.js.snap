// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Menu component Menu renderers 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`Testing requests init favorite 1`] = `
Array [
  Object {
    "payload": Object {
      "data": Array [
        Object {
          "star_1": false,
          "star_2": false,
          "star_3": false,
          "star_4": false,
          "star_5": false,
        },
      ],
    },
    "type": "init_favorite",
  },
]
`;

exports[`Testing requests save rating 1`] = `
Array [
  Object {
    "payload": Object {
      "data": Array [
        Object {
          "star_1": false,
          "star_2": false,
          "star_3": false,
          "star_4": false,
          "star_5": false,
        },
      ],
    },
    "type": "init_favorite",
  },
]
`;
